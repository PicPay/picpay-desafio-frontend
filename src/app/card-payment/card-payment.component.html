<div class="users">
    <div class="user" *ngFor="let user of users">
      <img src={{user.img}} />
      <div class="info-wrapper">
        <div class="user-name">
          {{ user.name}}
        </div>
  
        <div class="user-id-username">
          ID: {{user.id}} - Username: {{user.username}}
        </div>
      </div>
  
      <button type="button" class="pay-button" (click)="openModal(payByCard, user)">
        Pagar
      </button>
    </div>
</div>
  
  <ng-template class="modal-backdrop" #payByCard>
    <div class="modal">
      <div class="title">Pagamento para <span>{{user.username}}</span></div>
  
      <div class="content">
        <input type="number" placeholder="R$ 0,00" />
  
        <div class="cards">
          <mat-form-field>
            <mat-label>Escolha o cartão</mat-label>
            <mat-select>
              <mat-option *ngFor="let card of cards" [value]="card.card_number">
                Cartão com final {{ lastFourNumbers(card.card_number) }}
              </mat-option>
            </mat-select>
          </mat-form-field>
        </div>
  
        <button type="button" class="pay-button" (click)="paymentCard(user)">
          Pagar
        </button>
      </div>
    </div>
</ng-template>
