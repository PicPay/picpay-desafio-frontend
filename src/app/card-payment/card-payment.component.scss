/* List of Users */

.user {
    background: linear-gradient(186.5deg, #32cd32 0.83%, #134e13 100%);
    box-shadow: 0 1px 0 #ffffff, 0px -1px 0px #ffffff;
    color: #fff;
    padding: 20px 40px;
  
    display: flex;
    align-items: center;
  }

  .user-name{
    font-weight: bold;
  }

  .user > img {
    border-radius: 50%;
    height: 80px;
    width: 80px;
  }
  
  .user > .info-wrapper {
    margin-left: 16px;
  }
  
  .user > .pay-button {
    margin-left: auto;
  }

  
  /* Payment Modal */
  .modal-backdrop {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
  }
  
  .modal-backdrop.hidden {
    display: none;
  }
  
  .modal-backdrop:after {
    content: "";
    background: rgba(0, 0, 0, 0.5);
  }
  
  .modal {
    background: #fff;
    width: 420px;
  }

  ::ng-deep.mat-dialog-container{
    padding: 0px !important;
  }
  
  .modal > header > .title {
    background: #134e13;
    color: #ffffff;
    font-weight: 500;
    font-size: 16px;
    padding: 16px 24px;
  }
  
  .modal > header > .title > span {
    color: #7adc39;
  }
  
  .modal > .content {
    padding: 32px;
    text-align: center;
  }
  
  .modal > .content > input {
    border: 1px solid #ccc;
    border-radius: 6px;
    font-size: 20px;
    height: 48px;
    padding: 5px 20px;
  }
  
  .modal > .content > .cards > mat-select {
    background: #fff;
    border: 1px solid #ccc;
    border-radius: 6px;
    font-size: 18px;
    height: 48px;
    margin-top: 12px;
    padding: 10px 20px;
    width: 73%;
  }

  
  .modal > .content > button {
    border: 1px solid #ccc;
    border-radius: 6px;
    font-size: 20px;
    height: 48px;
    padding: 0 8px;
    margin-top: 24px;
  }
  
  .mat-raised-button.mat-button-disabled:not([class*="mat-elevation-z"]) {
    opacity: 60%;
  }