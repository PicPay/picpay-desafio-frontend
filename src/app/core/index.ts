export * from './components/index';
export * from './model/index';
export * from './services/index';
