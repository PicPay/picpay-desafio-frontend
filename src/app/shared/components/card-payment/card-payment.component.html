<div class="card" [class.card--hover]="isFullPage" [routerLink]="['/users', user.id]">
  <div *ngIf="!isFullPage" class="card__icon card__icon--withou-image">
    <span class="mdi mdi-currency-usd"></span>
  </div>
  <img *ngIf="isFullPage" class="card__icon" [src]="user.img" alt="User image" />
  <div class="card__info">
    <div>
      <strong>VocÃª</strong>
      <span> pagou </span>
      <strong>{{ user.name }}</strong>
    </div>
    <div class="card__date">
      <span>{{ payment.date | amTimeAgo }}</span>
    </div>
  </div>
  <div class="card__amount">
    <span>{{ payment.value | currency: 'BRL':true }}</span>
  </div>
</div>
