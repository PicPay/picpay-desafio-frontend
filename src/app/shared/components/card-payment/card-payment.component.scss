@import '~src/sass/variables';
@import '~src/sass/utils';

.card {
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: 10px;
    margin: 10px;
    padding: 10px;

    &--hover {
        cursor: pointer;

        &:hover {
            @include box-shadow();
        }
    }

    &__icon {
        width: 50px;
        height: 50px;
        border-radius: 100%;
        margin-right: 15px;

        &--withou-image {
            display: flex;
            align-items: center;
            justify-content: center;
            background: $primary;
            color: $light;
            font-size: 20px;
        }
    }

    &__image {
        width: 70px;
        height: 70px;
        border-radius: 10px;
        margin-right: 10px;
    }

    &__info {
        flex: 1;
        display: flex;
        flex-direction: column;
        white-space: nowrap;
        overflow: hidden;

        &--truncated {
            overflow: hidden;
            text-overflow: ellipsis;
        }
    }

    &__date {
        color: lighten($dark, 20%);
    }

    &__amount {
        color: $primary;
    }
}