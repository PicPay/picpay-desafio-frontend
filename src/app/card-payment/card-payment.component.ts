import { Component, OnInit, Inject } from '@angular/core';
import { CardPaymentService } from './service/card-payment.service';
import { MatDialog, MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';

@Component({
  selector: 'app-card-payment',
  templateUrl: './card-payment.component.html',
  styleUrls: ['./card-payment.component.scss']
})

export class CardPaymentComponent implements OnInit {

  users;
  user;

  formPay: FormGroup;

  cards = [
    {
      card_number: '1111111111111111',
      cvv: 789,
      expiry_date: '01/18',
    },
    {
      card_number: '4111111111111234',
      cvv: 123,
      expiry_date: '01/20',
    },
  ];
  
  constructor(
    private service: CardPaymentService, 
    public dialog: MatDialog,
    private fb: FormBuilder 
  ){ 
    this.formPay = this.fb.group({
      card_number: ['', Validators.required],
      value: ['', Validators.required],
      destination_user_id: ['']
    })
  }
  
  ngOnInit() {
    this.loadUsers();
  }

  loadUsers(){
    return this.service.userList().subscribe(data => {
      this.users = data;
    })
  }

  openDialog(template, row): void {
    this.user = row
    this.dialog.open(template);
  }

  lastFourNumbers(card){
    if(card) return card.substr(12);
  }

  paymentCard(){
    this.formPay.value.destination_user_id = this.user.id
    return this.service.payment(this.formPay.value).subscribe(data => {
      this.dialog.closeAll();     
      this.dialog.open(Alert, {
        data: {
          content: 'O pagamento foi concluido com sucesso.',
        },
      });
    }, err => {
      this.dialog.open(Alert, {
        data: {
          content: 'O pagamento n√£o foi concluido com sucesso.',
        },
      });
    })
  }
}

@Component({
  selector: 'alert',
  templateUrl: 'dialog/alert.component.html',
  styleUrls: ['dialog/alert.component.scss']
})
export class Alert {

  constructor(
    public dialogRef: MatDialogRef<Alert>,
    @Inject(MAT_DIALOG_DATA) public data: any) {}

}
