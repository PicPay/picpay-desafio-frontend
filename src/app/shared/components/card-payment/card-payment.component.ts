import { Component, Input, OnInit } from '@angular/core';
import { PaymentStorage } from '@shared/interfaces/payment-storage';
import { User } from '@shared/interfaces/user';
import { PicPayStore } from '@shared/stores/picpay.store';

@Component({
  selector: 'ngx-card-payment',
  templateUrl: 'card-payment.component.html',
  styleUrls: ['card-payment.component.scss'],
})
export class CardPaymentComponent implements OnInit {
  @Input() payment: PaymentStorage;
  @Input() isFullPage: boolean;

  user: User;

  constructor(private picPayStore: PicPayStore) {}

  ngOnInit(): void {
    this.picPayStore.loadUser(this.payment.destination_user_id).subscribe((user: User) => {
      this.user = user;
    });
  }
}
