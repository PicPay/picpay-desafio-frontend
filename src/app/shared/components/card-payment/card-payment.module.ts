import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { CardPaymentComponent } from './card-payment.component';
import { NgxSkeletonLoaderModule } from 'ngx-skeleton-loader';
import { MomentModule } from 'ngx-moment';

import 'moment/locale/pt-br';

@NgModule({
  declarations: [CardPaymentComponent],
  exports: [CardPaymentComponent],
  imports: [CommonModule, RouterModule, MomentModule, NgxSkeletonLoaderModule],
})
export class CardPaymentModule {}
