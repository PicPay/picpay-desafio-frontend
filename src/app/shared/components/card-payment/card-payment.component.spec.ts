import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { MockPicPayService } from '@mocks/MockPicPayService';
import { PicPayService } from '@services/picpay.service';
import { MomentModule } from 'ngx-moment';
import { NgxSkeletonLoaderModule } from 'ngx-skeleton-loader';

import { CardPaymentComponent } from './card-payment.component';

describe('CardPaymentComponent', () => {
  let component: CardPaymentComponent;
  let fixture: ComponentFixture<CardPaymentComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [CardPaymentComponent],
      imports: [RouterTestingModule, MomentModule, NgxSkeletonLoaderModule],
      providers: [{ provide: PicPayService, useClass: MockPicPayService }],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CardPaymentComponent);
    component = fixture.componentInstance;

    component.payment = {
      id: 1,
      date: new Date(),
      destination_user_id: 1,
      last_card_number: '1234',
      value: 100,
      comment: 'Test comment',
    };

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should create snapshot with payment', () => {
    expect(fixture).toMatchSnapshot('with_payment');
  });

  it('should create snapshot without payment', () => {
    component.payment = null;

    component.ngOnInit();
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot('without_payment');
  });

  it('test load user on init', () => {
    expect(component.user.id).toEqual(1);
    expect(component.user.name).toEqual('Test 1');
    expect(component.user.username).toEqual('@test1');
  });
});
