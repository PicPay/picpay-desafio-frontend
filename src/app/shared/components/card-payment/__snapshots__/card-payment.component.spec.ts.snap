// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPaymentComponent should create snapshot with payment: with_payment 1`] = `
<ngx-card-payment
  payment={[Function Object]}
  picPayStore={[Function PicPayStore]}
  user={[Function Object]}
>
  <div
    class="card"
    tabindex="0"
  >
    
    <div
      class="card__icon card__icon--withou-image"
    >
      <span
        class="mdi mdi-currency-usd"
      />
    </div>
    
    <div
      class="card__info"
    >
      <div>
        <strong>
          Test 1
        </strong>
      </div>
      
      <div
        class="card__comment"
      >
        <span>
          Test comment
        </span>
      </div>
      <div
        class="card__date"
      >
        <span>
          a few seconds ago
        </span>
      </div>
    </div>
    <div
      class="card__detail"
    >
      <span>
        **** 1234
      </span>
      <span
        class="card__amount"
      >
        R$100.00
      </span>
    </div>
  </div>
</ngx-card-payment>
`;

exports[`CardPaymentComponent should create snapshot without payment: without_payment 1`] = `
<ngx-card-payment
  payment="null"
  picPayStore={[Function PicPayStore]}
  user={[Function Object]}
>
  <div
    class="card"
  >
    <div
      class="card__icon"
    >
      <ngx-skeleton-loader>
        
        <span
          aria-busy="true"
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuetext="Loading..."
          class="loader progress"
          role="progressbar"
          style="width: 50px; height: 50px; margin: 0px 10px 0px 0px; border-radius: 100%;"
          tabindex="0"
        />
      </ngx-skeleton-loader>
    </div>
    <div
      class="card__info"
    >
      <div>
        <ngx-skeleton-loader>
          
          <span
            aria-busy="true"
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuetext="Loading..."
            class="loader progress"
            role="progressbar"
            style="width: 120px; margin: 0px;"
            tabindex="0"
          />
        </ngx-skeleton-loader>
      </div>
      <div>
        <ngx-skeleton-loader>
          
          <span
            aria-busy="true"
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuetext="Loading..."
            class="loader progress"
            role="progressbar"
            style="width: 100px; height: 15px; margin: 0px;"
            tabindex="0"
          />
        </ngx-skeleton-loader>
      </div>
    </div>
    <div
      class="card__detail"
    >
      <div>
        <ngx-skeleton-loader>
          
          <span
            aria-busy="true"
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuetext="Loading..."
            class="loader progress"
            role="progressbar"
            style="width: 70px; height: 15px;"
            tabindex="0"
          />
        </ngx-skeleton-loader>
      </div>
      <div>
        <ngx-skeleton-loader>
          
          <span
            aria-busy="true"
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuetext="Loading..."
            class="loader progress"
            role="progressbar"
            style="width: 60px; height: 15px; margin: 0px;"
            tabindex="0"
          />
        </ngx-skeleton-loader>
      </div>
    </div>
  </div>
</ngx-card-payment>
`;
